File: src/LanAudioRelay.Core/Networking/ControlServer.cs
using System.Net;
using System.Net.Sockets;
using LanAudioRelay.Core.Audio;
using LanAudioRelay.Core.Protocol;

namespace LanAudioRelay.Core.Networking;

public sealed class ControlServer : IAsyncDisposable
{
    private readonly Func<string> _pairingCodeProvider;
    private readonly object _gate = new();
    private CancellationTokenSource? _cts;
    private TcpListener? _listener;

    public ControlServer(Func<string> pairingCodeProvider)
    {
        _pairingCodeProvider = pairingCodeProvider;
    }

    public event EventHandler<PairingAcceptedEventArgs>? PairingAccepted;

    public Guid CurrentSessionId { get; private set; } = Guid.Empty;

    public void Start(int controlPort = AudioConstants.ControlPort)
    {
        if (_listener is not null)
        {
            throw new InvalidOperationException("The control server is already running.");
        }

        _cts = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, controlPort);
        _listener.Start();
        _ = Task.Run(() => AcceptLoopAsync(_cts.Token));
    }

    private async Task AcceptLoopAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested && _listener is not null)
        {
            try
            {
                var client = await _listener.AcceptTcpClientAsync(cancellationToken).ConfigureAwait(false);
                _ = Task.Run(() => HandleClientAsync(client, cancellationToken), cancellationToken);
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (ObjectDisposedException)
            {
                break;
            }
        }
    }

    private async Task HandleClientAsync(TcpClient client, CancellationToken cancellationToken)
    {
        using var _ = client;

        try
        {
            using var stream = client.GetStream();
            using var connection = new JsonLineConnection(stream);

            var hello = await connection.ReceiveAsync<HelloRequest>(cancellationToken).ConfigureAwait(false);
            if (hello.Type != "hello" || hello.ProtocolVersion != AudioConstants.ProtocolVersion)
            {
                await SendRejectedAsync(connection, "Unsupported protocol version.", cancellationToken).ConfigureAwait(false);
                return;
            }

            var serverNonce = Pairing.CreateNonce();
            await connection.SendAsync(
                new ChallengeResponse(
                    "challenge",
                    AudioConstants.ProtocolVersion,
                    Environment.MachineName,
                    serverNonce,
                    AudioSettings.Default),
                cancellationToken).ConfigureAwait(false);

            var pair = await connection.ReceiveAsync<PairRequest>(cancellationToken).ConfigureAwait(false);
            if (pair.Type != "pair" ||
                !Pairing.VerifyProof(_pairingCodeProvider(), hello.ClientNonce, serverNonce, pair.Proof))
            {
                await SendRejectedAsync(connection, "Invalid pairing code.", cancellationToken).ConfigureAwait(false);
                return;
            }

            var sessionId = Guid.NewGuid();
            lock (_gate)
            {
                CurrentSessionId = sessionId;
            }

            await connection.SendAsync(
                new PairResult(
                    "pairResult",
                    true,
                    null,
                    sessionId,
                    AudioConstants.MediaPort,
                    AudioSettings.Default),
                cancellationToken).ConfigureAwait(false);

            PairingAccepted?.Invoke(this, new PairingAcceptedEventArgs(sessionId, hello.ClientName));
        }
        catch
        {
            // A failed pairing attempt must not stop the receiver.
        }
    }

    private static Task SendRejectedAsync(
        JsonLineConnection connection,
        string reason,
        CancellationToken cancellationToken)
    {
        return connection.SendAsync(
            new PairResult(
                "pairResult",
                false,
                reason,
                Guid.Empty,
                AudioConstants.MediaPort,
                AudioSettings.Default),
            cancellationToken);
    }

    public async ValueTask DisposeAsync()
    {
        _cts?.Cancel();
        _listener?.Stop();
        _cts?.Dispose();
        await Task.CompletedTask.ConfigureAwait(false);
    }
}

public sealed class PairingAcceptedEventArgs : EventArgs
{
    public PairingAcceptedEventArgs(Guid sessionId, string clientName)
    {
        SessionId = sessionId;
        ClientName = clientName;
    }

    public Guid SessionId { get; }
    public string ClientName { get; }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMacCore/ControlServer.swift
import Foundation
import Network

public final class ControlServer: @unchecked Sendable {
    private let pairingCodeProvider: @Sendable () -> String
    private var listener: NWListener?
    private let queue = DispatchQueue(label: "LanAudioRelayMac.ControlServer")

    public var onPairingAccepted: (@Sendable (UUID, String) -> Void)?
    public private(set) var currentSessionId: UUID?

    public init(pairingCodeProvider: @escaping @Sendable () -> String) {
        self.pairingCodeProvider = pairingCodeProvider
    }

    public func start(port: UInt16 = AudioConstants.controlPort) throws {
        let listener = try NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: port)!)
        listener.newConnectionHandler = { [weak self] connection in
            guard let self else {
                connection.cancel()
                return
            }
            Task.detached {
                await self.handle(connection)
            }
        }
        listener.start(queue: queue)
        self.listener = listener
    }

    public func stop() {
        listener?.cancel()
        listener = nil
    }

    private func handle(_ nwConnection: NWConnection) async {
        let connection = TCPLineConnection(connection: nwConnection)
        do {
            try await connection.start()
            defer { connection.cancel() }

            let hello = try await connection.receive(HelloRequest.self)
            guard hello.type == "hello",
                  hello.protocolVersion == AudioConstants.protocolVersion
            else {
                try await sendRejected(connection, reason: "Unsupported protocol version.")
                return
            }

            let serverNonce = Pairing.createNonce()
            try await connection.send(ChallengeResponse(
                type: "challenge",
                protocolVersion: AudioConstants.protocolVersion,
                serverName: Host.current().localizedName ?? "Mac",
                serverNonce: serverNonce,
                audioSettings: .default
            ))

            let pair = try await connection.receive(PairRequest.self)
            guard pair.type == "pair",
                  Pairing.verifyProof(
                    expectedCode: pairingCodeProvider(),
                    clientNonce: hello.clientNonce,
                    serverNonce: serverNonce,
                    proof: pair.proof
                  )
            else {
                try await sendRejected(connection, reason: "Invalid pairing code.")
                return
            }

            let sessionId = UUID()
            currentSessionId = sessionId
            try await connection.send(PairResult(
                type: "pairResult",
                accepted: true,
                errorMessage: nil,
                sessionId: sessionId,
                mediaPort: Int(AudioConstants.mediaPort),
                audioSettings: .default
            ))
            onPairingAccepted?(sessionId, hello.clientName)
        } catch {
            connection.cancel()
        }
    }

    private func sendRejected(_ connection: TCPLineConnection, reason: String) async throws {
        try await connection.send(PairResult(
            type: "pairResult",
            accepted: false,
            errorMessage: reason,
            sessionId: UUID(uuidString: "00000000-0000-0000-0000-000000000000")!,
            mediaPort: Int(AudioConstants.mediaPort),
            audioSettings: .default
        ))
    }
}
